File: jvm-experiments-timeperiod/src/main/java/kr/kth/timeperiod/timerange/MinuteRange.java
package kr.kth.timeperiod.timerange;

import kr.kth.timeperiod.ITimeCalendar;
import kr.kth.timeperiod.ITimeFormatter;
import kr.kth.timeperiod.TimeCalendar;
import kr.kth.timeperiod.clock.ClockProxy;
import org.joda.time.DateTime;

import static kr.kth.commons.Guard.firstNotNull;

/**
 * 1분 단위 간격을 나타냅니다.
 * User: sunghyouk.bae@gmail.com
 * Date: 12. 12. 25.
 */
public class MinuteRange extends MinuteTimeRange {

    private static final long serialVersionUID = 9163926958263476312L;

    // region << Constructors >>

    public MinuteRange() {
        this(TimeCalendar.Default());
    }

    public MinuteRange(ITimeCalendar calendar) {
        this(ClockProxy.getClock().now(), calendar);
    }

    public MinuteRange(DateTime moment) {
        this(moment, TimeCalendar.Default());
    }

    public MinuteRange(DateTime moment, ITimeCalendar calendar) {
        super(moment, 1, calendar);
    }

    public MinuteRange(int year, int month, int day, int hour, int minute) {
        this(year, month, day, hour, minute, TimeCalendar.Default());
    }

    public MinuteRange(int year, int month, int day, int hour, int minute, ITimeCalendar calendar) {
        super(year, month, day, hour, minute, 1, calendar);
    }

    // endregion

    public int getYear() {
        return getStartYear();
    }

    public int getMonth() {
        return getStartMonth();
    }

    public int getDay() {
        return getStartDay();
    }

    public int getHour() {
        return getStartHour();
    }

    public int getMinute() {
        return getStartMinute();
    }

    public MinuteRange getPreviousMinute() {
        return addMinutes(-1);
    }

    public MinuteRange getNextMinute() {
        return addMinutes(1);
    }

    public MinuteRange addMinutes(int minutes) {
        DateTime start = getStart().withTime(getStartHour(), getStartMinute(), 0, 0);
        return new MinuteRange(start.plusMinutes(minutes), getTimeCalendar());
    }

    @Override
    protected String format(ITimeFormatter formatter) {
        ITimeFormatter fmt = firstNotNull(formatter, ITimeFormatter.getInstance());

        return fmt.getCalendarPeriod(fmt.getShortDate(getStart()),
                fmt.getShortTime(getStart()),
                fmt.getShortTime(getEnd()),
                getDuration());
    }
}

File: jvm-experiments-timeperiod/src/main/scala/kr/kth/timeperiod/timerange/MinuteRange.scala
package kr.kth.timeperiod.timerange

import kr.kth.commons.tools.ScalaHash
import kr.kth.timeperiod._
import org.joda.time.DateTime


abstract class MinuteTimeRange(start: DateTime, minuteCount: Int, timeCalendar: ITimeCalendar)
    extends CalendarTimeRange(start, start.plusMinutes(minuteCount), timeCalendar) {

    private val _minuteCount = minuteCount
    private val _endMinute = start.plusMinutes(minuteCount).getMinuteOfHour

    def this(year: Int, month: Int, day: Int, hour: Int, minute: Int, minuteCount: Int, calendar: ITimeCalendar) {
        this(new DateTime(year, month, day, hour, minute, 0), minuteCount, calendar)
    }

    def this(year: Int, month: Int, day: Int, hour: Int, minute: Int, minuteCount: Int) {
        this(year, month, day, hour, minute, minuteCount, TimeCalendar.Default)
    }

    def getMinuteCount: Int = _minuteCount

    override def getEndMinute: Int = _endMinute

    override def hashCode = ScalaHash.compute(super.hashCode, _minuteCount)

    protected override def buildStringHelper() =
        super.buildStringHelper()
        .add("minuteCount", _minuteCount)
        .add("endMinute", _endMinute)
}

/**
 * 분단위의 기간을 나타내는 클래스
 * User: sunghyouk.bae@gmail.com
 * Date: 13. 1. 17.
 */
class MinuteRange(moment: DateTime, calendar: ITimeCalendar)
    extends MinuteTimeRange(moment, 1, calendar) {

    def this(calendar: ITimeCalendar) {
        this(Clock.getClock.now, calendar)
    }

    def this() {
        this(TimeCalendar.Default)
    }

    def this(year: Int, month: Int, day: Int, hour: Int, minute: Int, calendar: ITimeCalendar) {
        this(new DateTime(year, month, day, hour, minute, 0, 0), calendar)
    }

    def this(year: Int, month: Int, day: Int, hour: Int, minute: Int) {
        this(year, month, day, hour, minute, TimeCalendar.Default)
    }

    def getYear: Int = getStartYear

    def getMonth: Int = getStartMonth

    def getDay: Int = getStartDay

    def getHour: Int = getStartHour

    def getMinute: Int = getStartMinute

    def getPreviousMinute: MinuteRange = addMinutes(-1)

    def getNextMinute: MinuteRange = addMinutes(1)

    def addMinutes(minutes: Int): MinuteRange = {
        val startMinute = getStart
                          .withTimeAtStartOfDay()
                          .withHourOfDay(getStartHour)
                          .withMinuteOfHour(getStartMinute)
        new MinuteRange(startMinute.plusMinutes(minutes), getTimeCalendar)
    }

    override protected def format(formatter: Option[ITimeFormatter]): String = {
        val fmt = formatter.getOrElse(TimeFormatter.instance)
        fmt.getCalendarPeriod(fmt.getShortDate(getStart),
            fmt.getShortTime(getStart),
            fmt.getShortTime(getEnd),
            getDuration)
    }
}

/**
 * 분단위의 기간 정보 ({@link MinuteRange}의 컬렉션
 */
class MinuteRangeCollection(moment: DateTime, minuteCount: Int, calendar: ITimeCalendar)
    extends MinuteTimeRange(moment, minuteCount, calendar) {

    def this(moment: DateTime, minuteCount: Int) {
        this(moment, minuteCount, TimeCalendar.Default)
    }

    def this(year: Int, month: Int, day: Int, hour: Int, minute: Int, minuteCount: Int, calendar: ITimeCalendar) {
        this(new DateTime(year, month, day, hour, minute, 0, 0), minuteCount, calendar)
    }

    def this(year: Int, month: Int, day: Int, hour: Int, minute: Int, minuteCount: Int) {
        this(year, month, day, hour, minute, minuteCount, TimeCalendar.Default)
    }

    def getMinutes: Iterable[MinuteRange] = {
        var startMinute = getStart
                          .withTimeAtStartOfDay()
                          .withHourOfDay(getStartHour)
                          .withMinuteOfHour(getStartMinute)

        (0 until getMinuteCount)
        .map(x => new MinuteRange(startMinute.plusMinutes(x), getTimeCalendar))
    }

}